File: client/packages/teampilot_search/test/file_index_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:teampilot_search/teampilot_search.dart';

void main() {
  final fixtureRoot = Directory('rust/tests/fixtures/file_index').absolute.path;

  test('builds and fuzzy queries indexed files', () async {
    final index = TpFileIndex();
    addTearDown(index.dispose);

    await index.build(fixtureRoot);
    final hits = index.query('router');

    expect(index.isBuilt, isTrue);
    expect(
      hits.map((hit) => hit.relativePath),
      contains('lib/app_router.dart'),
    );
    expect(
      hits.map((hit) => hit.relativePath),
      isNot(contains(contains('node_modules'))),
    );
  });

  test('queries files by contains mode and directories', () async {
    final index = TpFileIndex();
    addTearDown(index.dispose);

    await index.build(fixtureRoot);

    expect(
      index
          .query('chat', mode: TpFileMatchMode.contains)
          .map((hit) => hit.relativePath),
      contains('lib/chat_cubit.dart'),
    );
    expect(index.queryDirectories('lib'), contains('lib'));
  });

  test('returns large expanded query result sets without clipping', () async {
    final root = await Directory.systemTemp.createTemp('tp-file-index-');
    addTearDown(() => root.delete(recursive: true));
    final index = TpFileIndex();
    addTearDown(index.dispose);

    for (var i = 0; i < 600; i++) {
      await File(
        '${root.path}/expanded-${i.toString().padLeft(4, '0')}-${'x' * 120}.dart',
      ).writeAsString('');
    }

    await index.build(root.path);
    final hits = index.query(
      'expanded',
      mode: TpFileMatchMode.contains,
      limit: 100000,
    );

    expect(hits, hasLength(600));
    expect(index.truncated, isFalse);
  });
}

File: client/packages/teampilot_search/rust/tests/file_index_test.rs
use std::path::PathBuf;

use teampilot_search_rust::file_index::{FileIndex, FileIndexConfig, FileMatchMode};

fn fixture_root() -> String {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/fixtures/file_index")
        .to_string_lossy()
        .into_owned()
}

fn index(max_entries: u64) -> FileIndex {
    FileIndex::new(FileIndexConfig {
        root: fixture_root(),
        use_gitignore: true,
        max_entries,
    })
    .expect("fixture root should be readable")
}

#[test]
fn build_respects_gitignore_and_hidden() {
    let mut index = index(100);
    index.build().expect("index should build");

    assert_eq!(index.file_count(), 2);
    let paths: Vec<_> = index
        .query("", FileMatchMode::Contains, 10)
        .into_iter()
        .map(|hit| hit.relative_path)
        .collect();
    assert_eq!(paths, ["lib/app_router.dart", "lib/chat_cubit.dart"]);
}

#[test]
fn contains_mode_matches_basename() {
    let mut index = index(100);
    index.build().expect("index should build");

    let hits = index.query("ROUTER", FileMatchMode::Contains, 10);

    assert_eq!(hits.len(), 1);
    assert_eq!(hits[0].name, "app_router.dart");
    assert_eq!(hits[0].relative_path, "lib/app_router.dart");
}

#[test]
fn query_dirs_matches_basename() {
    let mut index = index(100);
    index.build().expect("index should build");

    assert_eq!(index.query_dirs("li", 10), ["lib"]);
}

#[test]
fn max_entries_truncates() {
    let mut index = index(1);
    index.build().expect("index should build");

    assert_eq!(index.file_count(), 1);
    assert!(index.truncated());
}
